packages/prio3/src/prio3.ts

```diff
@@ -319,184 +296,310 @@ export class Prio3<Measurement> implements Prio3Vdaf<Measurement> {
     }
   }
 
-  private decodePrepareMessage(encoded: Buffer): Buffer | null {
-    const { flp, prg } = this;
-
-    let jointRandCheck = null;
-    if (flp.jointRandLen > 0) {
-      [jointRandCheck, encoded] = split(encoded, prg.seedSize);
-    }
-
-    if (encoded.length > 0) {
-      throw new Error("unused bytes at end of prepare message share");
+  private decodePrepareMessage(encoded: Buffer): Buffer {
+    const { prg } = this;
+    if (encoded.length !== (this.useJointRand() ? prg.seedSize : 0)) {
+      throw new Error(
+        `expected prepare message to be ${prg.seedSize} bytes, but was ${encoded.length}`
+      );
     }
+    return encoded;
+  }
 
-    return jointRandCheck;
+  private async deriveSeed(
+    seed: Buffer,
+    usage: Usage,
+    parts: Uint8Array[]
+  ): Promise<Buffer> {
+    return Buffer.from(
+      await this.prg.deriveSeed(
+        seed,
+        this.domainSeparationTag(usage),
+        concat(parts)
+      )
+    );
   }
 
   private async buildLeaderShare(
     input: bigint[],
-    helperShares: Share[]
-  ): Promise<{
-    inputShare: bigint[];
-    blind: Buffer;
-    jointRandPart: Buffer;
-  }> {
+    helperShares: Share[],
+    nonce: Buffer,
+    rand: Buffer
+  ): Promise<Omit<Share, "proofShare" | "proofShareSeed">> {
     const { prg, field } = this;
-    const inputShare = helperShares.reduce(
-      (inputShare, helper) => field.vecSub(inputShare, helper.inputShare),
+    const measurementShare = helperShares.reduce(
+      (measurementShare, helper) =>
+        field.vecSub(measurementShare, helper.measurementShare),
       input
     );
-    const blind = Buffer.from(prg.randomSeed());
-    const encoded = field.encode(inputShare);
-    const jointRandPart = Buffer.from(
-      await prg.deriveSeed(
+    const measurementShareSeed = Buffer.from(
```

Comment:
Much appreciated — I was guessing on a lot of names. Which of those seems more descriptive?

Reply:
I think I prefer `wireMeasurementShare` -- we will also expand the seed into a full vector, and then encode that vector into a byte array to be used in joint randomness derivation. Using "wire" instead of "encoded" will hopefully distinguish the wire encoding of the share from that intermediate encoded byte string.